### PR TITLE
[SPARK-45844][SQL] Implement case-insensitivity for XML

### PR DESCRIPTION
**Pull request by shujingyang-db**
### What changes were proposed in this pull request?

This PR addresses the current lack of support for case-insensitive schema handling in XML file formats. Our approach now follows the `SQLConf` case insensitivity setting in both schema inference and file read operations.

We handle duplicate keys in the following behavior:
1. When we encounter duplicates (whether case-sensitive or not) in a row, we will convert them into an array and pick the first one we encounter as the array's name.
2. When we encounter duplicates across rows, we will also respect the first one we encounter 

Keys of the map-type data are string types and are not treated as field names, thereby not requiring case-sensitivity checks.

### Why are the changes needed?

To keep consistent with other file formats and reduce maintenance efforts.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Unit tests

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/xml/StaxXmlParser.scala [hunk 1]
```diff
@@ -274,7 +277,11 @@ class StaxXmlParser(
     val convertedValuesMap = collection.mutable.Map.empty[String, Any]
     val valuesMap = StaxXmlParserUtils.convertAttributesToValuesMap(attributes, options)
     valuesMap.foreach { case (f, v) =>
-      val nameToIndex = schema.map(_.name).zipWithIndex.toMap
+      val nameToIndex = if (caseSensitive) {
+        schema.map(_.name).zipWithIndex.toMap
+      } else {
+        CaseInsensitiveMap(schema.map(_.name).zipWithIndex.toMap)
+      }
```
[the remark this answers is not in the document]
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/xml/StaxXmlParser.scala:284]: +1

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/xml/XmlInferSchema.scala [hunk 2]
```diff
@@ -32,7 +32,12 @@ import scala.util.control.NonFatal
 import org.apache.spark.internal.Logging
 import org.apache.spark.rdd.RDD
 import org.apache.spark.sql.catalyst.expressions.ExprUtils
-import org.apache.spark.sql.catalyst.util.{DateFormatter, PermissiveMode, TimestampFormatter}
+import org.apache.spark.sql.catalyst.util.{
+  CaseInsensitiveMap,
+  DateFormatter,
+  PermissiveMode,
+  TimestampFormatter
+}
```
Comment (beliefer) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/xml/XmlInferSchema.scala:40]:
```suggestion
import org.apache.spark.sql.catalyst.util.{CaseInsensitiveMap, DateFormatter, PermissiveMode, TimestampFormatter}
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/xml/XmlInferSchema.scala [hunk 3]
```diff
@@ -198,30 +207,66 @@ private[sql] class XmlInferSchema(options: XmlOptions) extends Serializable with
    */
   private def inferObject(
       parser: XMLEventReader,
+      caseSensitive: Boolean,
       rootAttributes: Array[Attribute] = Array.empty): DataType = {
     val builder = ArrayBuffer[StructField]()
     val nameToDataType = collection.mutable.Map.empty[String, ArrayBuffer[DataType]]
+    // Initialize a map to hold field names with case sensitivity based on configuration.
+    // The map is only used in case *insensitive* mode
+    var fieldNames = if (caseSensitive) {
```
[the remark this answers is not in the document]
Comment (shujingyang-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/xml/XmlInferSchema.scala]: Can you please elaborate on this? I modified the map to make the usage more accurate.

[the remark this answers is not in the document]
Comment (shujingyang-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/xml/XmlInferSchema.scala]: We synced offline. I replaced it with the tree map. Thanks for the suggestion!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/xml/XmlInferSchema.scala [hunk 4]
```diff
@@ -50,7 +52,8 @@ private[sql] class XmlInferSchema(options: XmlOptions) extends Serializable with
     options.zoneId,
     options.locale,
     legacyFormat = FAST_DATE_FORMAT,
-    isParsing = true)
+    isParsing = true
+  )
```
Comment (beliefer) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/xml/XmlInferSchema.scala:51]: Please revert the change.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/xml/XmlInferSchema.scala [hunk 5]
```diff
@@ -97,20 +100,21 @@ private[sql] class XmlInferSchema(options: XmlOptions) extends Serializable with
...
       xml.sample(withReplacement = false, options.samplingRatio, 1)
     } else {
       xml
     }
     // perform schema inference on each row and merge afterwards
-    val rootType = schemaData.mapPartitions { iter =>
-      val xsdSchema = Option(options.rowValidationXSDPath).map(ValidatorUtil.getSchema)
+    val rootType = schemaData
+      .mapPartitions { iter =>
+        val xsdSchema = Option(options.rowValidationXSDPath).map(ValidatorUtil.getSchema)
```
Comment (beliefer) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/xml/XmlInferSchema.scala:107]: Please revert the three lines.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/xml/XmlInferSchema.scala [hunk 6]
```diff
@@ -97,20 +100,21 @@ private[sql] class XmlInferSchema(options: XmlOptions) extends Serializable with
...
 
-      iter.flatMap { xml =>
-        infer(xml, caseSensitive, xsdSchema)
+        iter.flatMap { xml =>
```
Comment (beliefer) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/xml/XmlInferSchema.scala:109]: Please revert this line.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/xml/XmlInferSchema.scala [hunk 7]
```diff
@@ -97,20 +100,21 @@ private[sql] class XmlInferSchema(options: XmlOptions) extends Serializable with
...
+          infer(xml, xsdSchema)
```
Comment (beliefer) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/xml/XmlInferSchema.scala:110]: two indent.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/xml/XmlInferSchema.scala [hunk 8]
```diff
@@ -115,7 +119,8 @@ private[sql] class XmlInferSchema(options: XmlOptions) extends Serializable with
     }
   }
 
-  def infer(xml: String,
+  def infer(
+      xml: String,
       xsdSchema: Option[Schema] = None): Option[DataType] = {
```
Comment (beliefer) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/xml/XmlInferSchema.scala:124]:
```suggestion
  def infer(xml: String, xsdSchema: Option[Schema] = None): Option[DataType] = {
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/xml/XmlInferSchema.scala [hunk 9]
```diff
@@ -200,7 +205,31 @@ private[sql] class XmlInferSchema(options: XmlOptions) extends Serializable with
...
+     *
+     * In case-sensitive mode: we will infer three fields: foo, Foo, FOO
+     * In case-insensitive mode, we will infer an array named by foo
+     * (as it's the first one we encounter)
+     */
+      val caseSensitivityOrdering: Ordering[String] = (x: String, y: String) =>
+        if (caseSensitive) {
+          x.compareTo(y)
+        } else {
+        x.compareToIgnoreCase(y)
```
Comment (sandip-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/xml/XmlInferSchema.scala]:
```suggestion
          x.compareToIgnoreCase(y)
```

**Review by sandip-db — COMMENTED** [review]
LGTM

File: sql/core/src/test/resources/test-data/xml-resources/attributes-case-sensitive.xml [hunk 10]
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0"?>
+<ROWSET>
+    <ROW>
+        <struct attr1="1">1</struct>
+        <array attr2="2">2</array>
+        <array Attr2="3">3</array>
+        <array aTTr2="4">4</array>
```
[the remark this answers is not in the document]
Comment (sandip-db) [reply sql/core/src/test/resources/test-data/xml-resources/attributes-case-sensitive.xml:7]: LGTM. Its a valid XML and intended to test case insentivity for attributes.

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.
